waiter/src/waiter/metrics_sync.clj

```diff
@@ -104,10 +106,63 @@
         (log/error "not registering request as it is missing request id")
         router-metrics-state))))
 
+(defn- clean-service-id->instance-id->metric
+  "Remove services from outer map that are not tracked by this router. Remove instances in the instance-id->metric map
+  that are not tracked by this router. This is done to prevent memory leaks in the service-id->instance-id->metric map"
+  [service-id-exists?-fn service-id-instance-id-active?-fn service-id->instance-id->metric]
+  (->> service-id->instance-id->metric
+       (utils/select-keys-pred service-id-exists?-fn)
+       keys
+       (pc/map-from-keys
+         (fn [service-id]
+           (let [instance-id->metric (get service-id->instance-id->metric service-id)]
+             (utils/select-keys-pred
+               (fn [instance-id]
+                 (service-id-instance-id-active?-fn service-id instance-id))
+               instance-id->metric))))))
+
+(defn- merge-instance-id->metric-maps
+  "Merges two instance-id->metric maps iterating through the instance-ids in both maps and checking which map has the
+  latest metric based on 'updated-at'. If the metric does not exist in one of the maps, then defaults to the non nil
+  metric."
+  [instance-id->metric-1 instance-id->metric-2]
+  (pc/map-from-keys
```

Comment:
We should use `merge-with` here.
https://clojuredocs.org/clojure.core/merge-with